src/EventStore.Core/Services/Transport/Grpc/Enumerators.StreamSubscription.cs

```diff
@@ -42,105 +53,447 @@ public class StreamSubscription : IAsyncEnumerator<ResolvedEvent> {
+						_current = current;
+						return true;
+					}
+
+					var correlationId = Guid.NewGuid();
+
+					var readNextSource = new TaskCompletionSource<bool>();
 
-				goto ReadLoop;
+					Log.Trace(
+						"Catch-up subscription {_subscriptionId} to {_streamName} reading next page starting from {_nextRevision}",
+						_subscriptionId, _streamName, _nextRevision);
 
-				void OnMessage(Message message) {
-					if (message is ClientMessage.NotHandled notHandled &&
-					    RpcExceptions.TryHandleNotHandled(notHandled, out var ex)) {
-						readNextSource.TrySetException(ex);
-						return;
+					_bus.Publish(new ClientMessage.ReadStreamEventsForward(
+						correlationId, correlationId, new CallbackEnvelope(OnMessage), _streamName,
+						_nextRevision.ToInt64(), 32, _resolveLinks, false, default, _user));
+
+					var isEnd = await readNextSource.Task.ConfigureAwait(false);
+
+					if (_buffer.TryDequeue(out current)) {
+						_current = current;
+						return true;
 					}
 
-					if (!(message is ClientMessage.ReadStreamEventsForwardCompleted completed)) {
-						readNextSource.TrySetException(
-							RpcExceptions.UnknownMessage<ClientMessage.ReadStreamEventsForwardCompleted>(message));
-						return;
+					if (isEnd) {
+						return false;
 					}
 
-					switch (completed.Result) {
-						case ReadStreamResult.Success:
-							foreach (var @event in completed.Events) {
-								_buffer.Enqueue(@event);
-							}
+					if (_disposedTokenSource.IsCancellationRequested) {
+						return false;
+					}
 
-							_nextRevision = StreamRevision.FromInt64(completed.NextEventNumber);
-							readNextSource.TrySetResult(true);
+					goto ReadLoop;
+
+					void OnMessage(Message message) {
+						if (message is ClientMessage.NotHandled notHandled &&
+						    RpcExceptions.TryHandleNotHandled(notHandled, out var ex)) {
+							readNextSource.TrySetException(ex);
 							return;
-						case ReadStreamResult.NoStream:
-							readNextSource.TrySetException(RpcExceptions.StreamNotFound(_streamName));
+						}
+
+						if (!(message is ClientMessage.ReadStreamEventsForwardCompleted completed)) {
+							readNextSource.TrySetException(
+								RpcExceptions.UnknownMessage<ClientMessage.ReadStreamEventsForwardCompleted>(message));
 							return;
-						case ReadStreamResult.StreamDeleted:
-							readNextSource.TrySetException(RpcExceptions.StreamDeleted(_streamName));
+						}
+
+						switch (completed.Result) {
+							case ReadStreamResult.Success:
+								foreach (var @event in completed.Events) {
+									if (StreamRevision.FromInt64(@event.OriginalEvent.EventNumber) < _startRevision) {
+										continue;
+									}
+
+									_buffer.Enqueue(@event);
+								}
+
+								_nextRevision = StreamRevision.FromInt64(completed.NextEventNumber);
+								readNextSource.TrySetResult(completed.IsEndOfStream);
+								return;
+							case ReadStreamResult.NoStream:
+								readNextSource.TrySetException(RpcExceptions.StreamNotFound(_streamName));
```

Comment:
Should we not start a subscription if there is no stream?

Reply:
This is the current behavior from the client api